### PR TITLE
Refactor circleci to reuse commands from precommit

### PR DESCRIPTION
**Pull request by dan-pandori**
## Description

Run various hooks in circleci via calling them through precommit. This should reduce the likelihood that there is a divergence between the two.
https://github.com/HumanCompatibleAI/imitation/issues/605

It's somewhat difficult to pass the arguments as we'd like to, due to:
https://github.com/pre-commit/pre-commit/issues/2083

To get around this, we directly rely on NUM_CPUS in the bash call to pytype. For flake8, we instead allow the existing flake8 parallelism to work w/o specifying --j.

## Testing

Confirmed that circleci continues to pass, and ran the hooks locally.

### CODE REVIEW

**Comment by dan-pandori** [comment]
@AdamGleave, ready for review when you've got a chance.

File: .circleci/config.yml [hunk 1]
```diff
@@ -231,23 +231,40 @@ jobs:
...
-          command: codespell -I .codespell.skip --skip='*.pyc,tests/testdata/*,*.ipynb,*.csv' ${SRC_FILES}
+          command: pre-commit run --all codespell
+
+      - run:
+          name: lint-misc
+          command: |
+            pre-commit run --all check-ast
+            pre-commit run --all trailing-whitespace
+            pre-commit run --all end-of-file-fixer
+            pre-commit run --all check-toml
```
Comment (AdamGleave) [inline .circleci/config.yml:267]: Should we have `check-added-large-files` as well?

Maybe that's OK to exclude as won't cause CI to fail on subsequent commits.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -42,7 +42,7 @@ repos:
   rev: v2.2.2
   hooks:
   - id: codespell
-    args: ["--skip=*.pyc,tests/testdata/*,*.ipynb,*.csv"]
+    args: ["--skip=*.pyc,tests/testdata/*,*.ipynb,*.csv","--ignore-words-list=reacher"]
```
Comment (AdamGleave) [inline .pre-commit-config.yaml:45]: We used to have `iff` skipped in `.codespell.skip` -- is that not needed?

We should probably also delete `.codespell.skip` if no longer used.

File: .circleci/config.yml [hunk 3]
```diff
@@ -231,23 +231,40 @@ jobs:
```
Comment (AdamGleave) [inline .circleci/config.yml:231]: Should we replace the shellcheck part of this file with the pre-commit hook?

File: .circleci/config.yml [hunk 4]
```diff
@@ -231,23 +231,40 @@ jobs:
...
+
+      - run:
+          name: lint-misc
+          command: |
+            pre-commit run --all check-ast
+            pre-commit run --all trailing-whitespace
+            pre-commit run --all end-of-file-fixer
+            pre-commit run --all check-toml
 
   doctest:
```
Comment (AdamGleave) [inline .circleci/config.yml:269]: Should we replace this with `pre-commit run --all docs`?

**Review by AdamGleave — CHANGES_REQUESTED** [review]
Left some minor comments inline that should be addressed before merging -- couple of hooks that were missed (or if not included on purpose, we should add a comment to explain it).

Other question is higher-level. `pre-commit` installs these hooks in their own virtual environment. I think our caching logic doesn't handle this as the CircleCI logs suggest it's installing black, etc from scratch each time. It's quite quick (the dominant cost is actually doing the type checking and running darglint, which is very expensive) but we should probably still update the caching logic to save these extra directories.

**Comment by AdamGleave** [comment]
@anshradh would you mind finishing this PR off now that Dan's finished his trial (I just had a few minor suggestions for changes)?

**Comment by anshradh** [comment]
Yup can do!

File: .circleci/config.yml (hunk 4 above)
Comment (anshradh) [reply .circleci/config.yml:257]: Ignoring per our earlier conversation.

File: .circleci/config.yml (hunk 3 above)
Comment (anshradh) [reply .circleci/config.yml:218]: I tried to do this, not confident I got it right though.

File: .pre-commit-config.yaml (hunk 2 above)
Comment (anshradh) [reply .pre-commit-config.yaml]: Done.

File: .circleci/config.yml (hunk 1 above)
Comment (anshradh) [reply .circleci/config.yml:254]: Added.

File: .circleci/config.yml (hunk 3 above)
Comment (AdamGleave) [reply .circleci/config.yml:218]: One small issue: `types: [bash]` actually skips everything (even with `pre-commit run --all shellcheck` nothing gets run). Leaving it as the default `types: [shell]` seems to work. My guess is pre-commit isn't smart enough to parse the shebang and realize we're using bash not sh, and is just going off of file extension.

I've pushed a commit to fix this.

**Comment by AdamGleave** [comment]
Oh, doh, seems like the shellcheck hook needs Docker to run and we can't do Docker-in-Docker on CircleCI. So might have to just stick with the orb anyway.

**Review by AdamGleave — APPROVED** [review]
LGTM. Will merge once tests pass.

**Comment by Rocamonde** [comment]
Seems like linting is failing after this has been merged.

**Comment by AdamGleave** [comment]
Yes there's one notebook it wants to reformat. Can you open a PR to fix
this? I don't know why the tests passed before merge given this.

**Pull request merged**